### PR TITLE
[12.0][ADD] base_phone_rate_import_bandwith

### PR DESCRIPTION
**Pull request by murtuzasaleh**
WIP

### CODE REVIEW

**Comment by murtuzasaleh** [comment]
@dreispt

File: base_phone_rate_import_bandwith/README.rst [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    :target: https://github.com/max3903
+    :alt: max3903
+
+Current `maintainers <https://odoo-community.org/page/maintainer-role>`__:
+
+|maintainer-wolfhall| |maintainer-max3903| 
+
+This module is part of the `OCA/vertical-isp <https://github.com/OCA/vertical-isp/tree/12.0/base_phone_rate_import_bandwith>`_ project on GitHub.
+
+You are welcome to contribute. To learn how please visit https://odoo-community.org/page/Contribute.
```
Comment (dreispt) [inline base_phone_rate_import_bandwith/README.rst]: It's easier to use the `readme` fragments subdirectory, and put here a README placeholder file.

File: base_phone_rate_import_bandwith/readme/CONFIGURE.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+This module needs no additional configuration post installation
```
Comment (dreispt) [inline base_phone_rate_import_bandwith/readme/CONFIGURE.rst:1]: Remember to add USAGE.rst - it is required.

File: base_phone_rate_import_bandwith/wizard/base_phone_rate_import.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            if not data.upload_file:
+                raise Warning(_('Please select a file to proceed.'))
+            if not filename_str[-5:] == ".xlsx":
+                raise Warning(_('Select .xlsx file only'))
+            csv_data = base64.decodestring(data.upload_file)
+            temp_path = tempfile.gettempdir()
+            fp = open(temp_path + '/xsl_file.xlsx', 'wb+')
+            fp.write(csv_data)
+            fp.close()
+            wb = open_workbook(temp_path + '/xsl_file.xlsx')
```
Comment (dreispt) [inline base_phone_rate_import_bandwith/wizard/base_phone_rate_import.py]: You should be able, without having to write it to a temp disk file.
Not 100% sure if xlrd supports that, though.

File: base_phone_rate_import_bandwith/wizard/base_phone_rate_import.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            for sheet in wb.sheets():
+                for rownum in range(0, sheet.nrows):
+                    if rownum > 0:
+                        data_list.append(sheet.row_values(rownum))
+            for values in data_list:
```
Comment (dreispt) [inline base_phone_rate_import_bandwith/wizard/base_phone_rate_import.py]: Not a problem, but I would have used a single loop over the lines...

File: base_phone_rate_import_bandwith/wizard/base_phone_rate_import.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                name = values[0]
+                if ' (' in values[0]:
+                    country = name.split(' (')
+                    country_name = country[0]
+                else:
+                    country_name = values[0]
+                country_rec = res_country_obj.search([('name', '=',
```
Comment (dreispt) [inline base_phone_rate_import_bandwith/wizard/base_phone_rate_import.py]: More robust if you ignore letter case:
```suggestion
                country_rec = res_country_obj.search([('name', '=ilike',
```

File: base_phone_rate_import_bandwith/wizard/base_phone_rate_import.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+                            country = name.split(' (')
+                            country_name = country[0]
+                        else:
+                            country_name = name
+                        country_rec = res_country_obj.search(
+                            [('name', '=ilike', country_name)])
+                        phone_rate_obj.create({'name': name,
+                                               'country_id': country_rec.id,
+                                               'dial_prefix': bandwith_data[1],
+                                               'rate': bandwith_data[2]})
```
Comment (dreispt) [inline base_phone_rate_import_bandwith/wizard/base_phone_rate_import.py]: You must update an existing rate if the dial prefix already exists.

**Comment by murtuzasaleh** [comment]
@dreispt @max3903 Can you please review this PR.

**Comment by dreispt** [comment]
Ping @kvedder-amplex

File: base_phone_rate_import_bandwith/wizard/base_phone_rate_import.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        phone_rate_obj = self.env['phone.rate']
+        for data in self:
+            filename_str = tools.ustr(data.filename)
+            if not data.upload_file:
+                raise Warning(_('Please select a file to proceed.'))
+            if not filename_str[-5:] == ".xlsx":
+                raise Warning(_('Select .xlsx file only'))
+            csv_data = base64.decodestring(data.upload_file)
+            temp_path = tempfile.gettempdir()
+            fname = temp_path + '/xsl_file.xlsx'
```
Comment (dreispt) [inline base_phone_rate_import_bandwith/wizard/base_phone_rate_import.py:33]: Nitpick: user `os.path.join`

File: base_phone_rate_import_bandwith/wizard/base_phone_rate_import.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+            fp = open(fname, 'wb+')
```
Comment (dreispt) [inline base_phone_rate_import_bandwith/wizard/base_phone_rate_import.py:34]: I would rather prefer to use an in memory IOStream, but I won't pick on this.

File: base_phone_rate_import_bandwith/wizard/base_phone_rate_import.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+            fp.write(csv_data)
+            fp.close()
+            wb = open_workbook(fname)
+            for sheet in wb.sheets():
+                for rownum in range(0, sheet.nrows):
+                    if rownum > 0:
```
Comment (dreispt) [inline base_phone_rate_import_bandwith/wizard/base_phone_rate_import.py:40]: This `if` is not necessary if you use `range(1, ...)`

File: base_phone_rate_import_bandwith/wizard/base_phone_rate_import.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+                        name = bandwith_data[0]
+                        if ' (' in bandwith_data[0]:
+                            country = name.split(' (')
+                            country_name = country[0]
+                        else:
+                            country_name = name
+                        country_rec = res_country_obj.search(
+                            [('name', '=ilike', country_name)])
+                        phone_rate_rec = phone_rate_obj.search(
+                            [('dial_prefix', '=', bandwith_data[1])])
```
Comment (dreispt) [inline base_phone_rate_import_bandwith/wizard/base_phone_rate_import.py:51]: `bandwith_data[1]` returns a float and we get a zero decimal when it is forced as a char: `92.0`
We don't want that, and need a proper conversion, to get `92` instead.
This needs fixing.

File: base_phone_rate_import_bandwith/wizard/base_phone_rate_import.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+                        if phone_rate_rec:
+                            phone_rate_rec.rate = bandwith_data[2]
+                        else:
+                            phone_rate_obj.create(
+                                {'name': name,
+                                 'country_id': country_rec.id,
+                                 'dial_prefix': bandwith_data[1],
+                                 'rate': bandwith_data[2]})
+            os.remove(fname)
```
Comment (dreispt) [inline base_phone_rate_import_bandwith/wizard/base_phone_rate_import.py:60]: :+1:

**Review by dreispt — CHANGES_REQUESTED** [review]

**Comment by dreispt** [comment]
Superseded by #45 Thank you @murtuzasaleh

**Pull request closed without merging**